R/utils.R

```diff
@@ -10,9 +10,56 @@ print.rtoot_bearer <- function(x,...){
   invisible(x)
 }
 
+# prepare urls
 prepare_url <- function(instance){
   url <- httr::parse_url("")
   url$hostname <- instance
   url$scheme <- "https"
   url
 }
+
+# process the header of a get request
+parse_header <- function(header){
+  # rate limit fields should always be present but the link field is not always there
+  df <- tibble::tibble(rate_limit=header[["x-ratelimit-limit"]],
+                       rate_remaining=header[["x-ratelimit-remaining"]],
+                       rate_reset=format_date(header[["x-ratelimit-reset"]]))
+  if("link"%in%names(header)){
+    vars_to_search <- c("max_id","min_id","since_id")
+    links <- regmatches(header[["link"]], gregexpr("https[^>]+", header[["link"]]))[[1]]
+    query_params <- lapply(links,function(x){
+      query <- httr::parse_url(x)[["query"]]
+      query <- query[names(query)%in%vars_to_search]
+    })
+    page_param <- dplyr::bind_cols(query_params)
+
+    return(dplyr::bind_cols(df,page_param))
+  } else{
+    return(df)
+  }
+}
+
+process_request <- function(token = NULL,
+                path,
+                instance = NULL,
+                params,
+                anonymous = FALSE,
+                parse = TRUE,
+                FUN = identity
+){
+  output <- make_get_request(token = token,path = path,
+                             instance = instance, params = params,
+                             anonymous = anonymous)
+  if (isTRUE(parse)) {
+    header <- attr(output,"headers")
+    if(length(output)<1){
+      output <- tibble::tibble()
+    } else if(length(output[[1]])==1){ #TODO:fragile?
```

Comment:
This is indeed a bit fragile.

Reply:
Yeah trying to work out a better solution. Maybe check for depth of list?
```r
# https://stackoverflow.com/a/13433689
depth <- function(this,thisdepth=0){
  if(!is.list(this)){
    return(thisdepth)
  }else{
    return(max(unlist(lapply(this,depth,thisdepth=thisdepth+1))))    
  }
} 
```
if depth =1 then we have only one object and if depth>1 we have a list of object?

Reply:
I think a better way to handle this is to use some crazy functional programming technique.

```r
## vectorizing
v <- function(FUN) {
    v_FUN <- function(x) {
        dplyr::bind_rows(lapply(x, FUN))
    }
    return(v_FUN)
}
```

For the functions that need vectoring, pass `v(parse_status)` as `FUN` to `process_request`

Reply:
Holy! I try it out

Reply:
but does it capture cases where we would usually expect several objects returned, but we only get one?

Reply:
I think it applies only to `get_context`. For that case, we just need to craft a specialize `FUN` for it.

So:

| Function | FUN |
|-------------|--------|
| `get_status` | parse_status |
| `get_public_timeline` | v(parse_status) |
| `get_context` | a specialize FUN |

Reply:
I test it

Reply:
https://github.com/schochastics/rtoot/blob/89bed02988df595e3c767b661805450c39220b49/R/timelines_statuses.R#L97-L100

Something like

```r
parse_context <- function(output) {
    temp_output <- list()
    temp_output$ancestors <- dplyr::bind_rows(lapply(output$ancestors, parse_status))
    temp_output$descendants <- dplyr::bind_rows(lapply(output$descendants, parse_status))
     temp_output
}
```